### ISSUE TITLE
Project level 4 event class 

### ISSUE THREAD

**Issue by ARPspoofing** [nus-cs2030/2021-s2#220]
# Summary
Issue with project level 4
## Description 
Hello! Can any kind soul explain to me how do I go about attempting level 4? I know that we need to create a new event class but how do I make the server go from resting to ready again after the resting time? Do I need to check the subsequent scanner arrival times and make the server stop resting when the conditions are met? Thanks for the help guys!

## Screenshots (if any): 
Insert Images here if necessary

**Comment by yuyitan** [comment]
The way I solved Level 4 was with the addition of 2 new event types, ServerRestEvent and ServerBackEvent. ServerRestEvent follows after each DoneEvent and when this event is triggered, the simulator checks if the server has to rest and add the resting time to the server's availableTime as well as modify a boolean attribute to indicate that the server is resting. The ServerBackEvent may not be necessary but I used it to ensure that my server is no longer resting.

**Comment by yaohua98** [comment]
> The way I solved Level 4 was with the addition of 2 new event types, ServerRestEvent and ServerBackEvent. ServerRestEvent follows after each DoneEvent and when this event is triggered, the simulator checks if the server has to rest and add the resting time to the server's availableTime as well as modify a boolean attribute to indicate that the server is resting. The ServerBackEvent may not be necessary but I used it to ensure that my server is no longer resting.
> 
> Edit: Issue #120 may be of help!

but the problem i had was that some wait events that are generated before the rest time is added have timestamp that does not take into account of the rest time, i think this is because the rest time is added to the server only when the rest event is polled for my case. Did you update the server times differently?

**Comment by yuyitan** [comment]
> > The way I solved Level 4 was with the addition of 2 new event types, ServerRestEvent and ServerBackEvent. ServerRestEvent follows after each DoneEvent and when this event is triggered, the simulator checks if the server has to rest and add the resting time to the server's availableTime as well as modify a boolean attribute to indicate that the server is resting. The ServerBackEvent may not be necessary but I used it to ensure that my server is no longer resting.
> > Edit: Issue #120 may be of help!
> 
> but the problem i had was that some wait events that are generated before the rest time is added have timestamp that does not take into account of the rest time, i think this is because the rest time is added to the server only when the rest event is polled for my case. Did you update the server times differently?

Hi, I encountered a similar problem with pre-scheduled events in the past too (Issue #159). I had to do some additional checking using peek() to check if I need to reschedule certain events should the server be resting during that time, followed by remove() if that event is found to have a wrong timing. Btw, servers can still take in waiting customers even if they are resting, so the main concern will be to check whether the ServeEvent has to be delayed to take into account the server rest length.

**Comment by ARPspoofing** [comment]
> > > The way I solved Level 4 was with the addition of 2 new event types, ServerRestEvent and ServerBackEvent. ServerRestEvent follows after each DoneEvent and when this event is triggered, the simulator checks if the server has to rest and add the resting time to the server's availableTime as well as modify a boolean attribute to indicate that the server is resting. The ServerBackEvent may not be necessary but I used it to ensure that my server is no longer resting.
> > > Edit: Issue #120 may be of help!
> > 
> > 
> > but the problem i had was that some wait events that are generated before the rest time is added have timestamp that does not take into account of the rest time, i think this is because the rest time is added to the server only when the rest event is polled for my case. Did you update the server times differently?
> 
> Hi, I encountered a similar problem with pre-scheduled events in the past too (Issue #159). I had to do some additional checking using peek() to check if I need to reschedule certain events should the server be resting during that time, followed by remove() if that event is found to have a wrong timing. Btw, servers can still take in waiting customers even if they are resting, so the main concern will be to check whether the ServeEvent has to be delayed to take into account the server rest length.

@yuyitan hi! can you elaborate what do you mean by remove() if that event is found to have wrong timing? Why would the event have wrong timing if the serverRestEvent works properly?

**Comment by Xeph-o** [comment]
What I did was to have a RestEvent, in which I input the resttime via the nextEvent(time) in my main class. By doing so, my next serve event is only be triggered by the done / rest event, and not by the wait event. The wait event merely adds my customer to the waiting queue.

**Comment by yuyitan** [comment]
> > > > The way I solved Level 4 was with the addition of 2 new event types, ServerRestEvent and ServerBackEvent. ServerRestEvent follows after each DoneEvent and when this event is triggered, the simulator checks if the server has to rest and add the resting time to the server's availableTime as well as modify a boolean attribute to indicate that the server is resting. The ServerBackEvent may not be necessary but I used it to ensure that my server is no longer resting.
> > > > Edit: Issue #120 may be of help!
> > > 
> > > 
> > > but the problem i had was that some wait events that are generated before the rest time is added have timestamp that does not take into account of the rest time, i think this is because the rest time is added to the server only when the rest event is polled for my case. Did you update the server times differently?
> > 
> > 
> > Hi, I encountered a similar problem with pre-scheduled events in the past too (Issue #159). I had to do some additional checking using peek() to check if I need to reschedule certain events should the server be resting during that time, followed by remove() if that event is found to have a wrong timing. Btw, servers can still take in waiting customers even if they are resting, so the main concern will be to check whether the ServeEvent has to be delayed to take into account the server rest length.
> 
> @yuyitan hi! can you elaborate what do you mean by remove() if that event is found to have wrong timing? Why would the event have wrong timing if the serverRestEvent works properly?

Hi, my reply was for yaohua98! Whether this problem may occur depends on your code and program design. For me, it was a problem because some of my serveEvent that were scheduled based on the server's availTime became invalid halfway throughout the runTime simulation (e.g. halfway server needs to rest -- this is when the server availTime is updated). Since the server went to rest, the previously scheduled serveEvent would need to be rescheduled to a later time when the server is done resting.

**Comment by dumbasssoul** [comment]
> What I did was to have a RestEvent, in which I input the resttime via the nextEvent(time) in my main class. By doing so, my next serve event is only be triggered by the done / rest event, and not by the wait event. The wait event merely adds my customer to the waiting queue.

is there a way to follow the original sequence of arrive -> wait/serve -> done -> rest -> back (end) ?

**Comment by NIL-99** [comment]
My implementation for level 4 will be to create a RestEvent (additional event) which will be called from a DoneEvent. An eventTime from DoneEvent will be passed onto the RestEvent in which, the server will extract the next rest timing/duration from the list of rest timing inputs. The RestEvent will not return any other event for my scenario

**Comment by AdelineWongg** [comment]
> My implementation for level 4 will be to create a RestEvent (additional event) which will be called from a DoneEvent. An eventTime from DoneEvent will be passed onto the RestEvent in which, the server will extract the next rest timing/duration from the list of rest timing inputs. The RestEvent will not return any other event for my scenario

@NIL-99 then how do u settle the Customers waiting in queue?

**Comment by lohwenxin** [comment]
> What I did was to have a RestEvent, in which I input the resttime via the nextEvent(time) in my main class. By doing so, my next serve event is only be triggered by the done / rest event, and not by the wait event. The wait event merely adds my customer to the waiting queue.

hello, I did this too but it says that there are cyclic dependencies. May I know how you got rid of the cyclic if you had that problem in your code?

**Comment by yuyitan** [comment]
> @yuyitan Hello! May I know how does this work flow help to settle "Wait"? because wait's next event is new Serve( .., time of ServeEvent,..) which will be put into the pq.
> After server rest and if they have someone in the queue, how do they go back to serve them?( because if Rest -> Serve directly, will cause cyclic dependencies).

Hi, sorry I can't advise much on this as I performed some checking for the nextEvent() method in my Simulator class when processing the PQ. This includes checking whether to queue a serveEvent. For Level 5, I tweaked it such that I will only queue a serveEvent from a waitEvent if there is only 1 customer in queue and serverRest does not exist. Subsequently, whenever a server comes back from rest (serverBackEvent), I will check the server internal queue and queue the next customer waiting in line (if any) for serving. I gotta admit that my code doesn't follow the OOP principles strictly, but desperate times desperate measures :")
